test/cpp/api/tensor.cpp

```diff
@@ -304,27 +478,81 @@ TEST(TensorTest, MultidimTensorCtor_CUDA) {
   }
 }
 
-TEST(TensorTest, PrettyPrintInitListTensor) {
+TEST(TensorTest, TensorCtorPreservesInitListDtype) {
+  ASSERT_EQ(torch::tensor({1, 2, 3}).dtype(), torch::kInt);
+  ASSERT_EQ(torch::tensor({{1, 2, 3}}).dtype(), torch::kInt);
+  ASSERT_EQ(torch::tensor({1., 2., 3.}).dtype(), torch::kDouble);
+  ASSERT_EQ(torch::tensor({{1., 2., 3.}}).dtype(), torch::kDouble);
+
+  ASSERT_EQ(torch::tensor({1, 2, 3}, torch::TensorOptions()).dtype(), torch::kInt);
+  ASSERT_EQ(torch::tensor({{1, 2, 3}}, torch::TensorOptions()).dtype(), torch::kInt);
+  ASSERT_EQ(torch::tensor({1., 2., 3.}, torch::TensorOptions()).dtype(), torch::kDouble);
+  ASSERT_EQ(torch::tensor({{1., 2., 3.}}, torch::TensorOptions()).dtype(), torch::kDouble);
+}
+
+TEST(TensorTest, PrettyPrintTensorDataContainer) {
   {
     ASSERT_EQ(
-      c10::str(torch::detail::InitListTensor(1.1)),
+      c10::str(torch::detail::TensorDataContainer(1.1)),
       "1.1");
   }
   {
     ASSERT_EQ(
-      c10::str(torch::detail::InitListTensor({1.1, 2.2})),
+      c10::str(torch::detail::TensorDataContainer({1.1, 2.2})),
       "{1.1, 2.2}");
   }
   {
     ASSERT_EQ(
-      c10::str(torch::detail::InitListTensor({{1, 2}, {3, 4}})),
+      c10::str(torch::detail::TensorDataContainer({{1, 2}, {3, 4}})),
       "{{1, 2}, {3, 4}}");
   }
   {
     ASSERT_EQ(
-      c10::str(torch::detail::InitListTensor({{{{{{{{1.1, 2.2, 3.3}}}}}, {{{{{4.4, 5.5, 6.6}}}}}, {{{{{7.7, 8.8, 9.9}}}}}}}})),
+      c10::str(torch::detail::TensorDataContainer({{{{{{{{1.1, 2.2, 3.3}}}}}, {{{{{4.4, 5.5, 6.6}}}}}, {{{{{7.7, 8.8, 9.9}}}}}}}})),
       "{{{{{{{{1.1, 2.2, 3.3}}}}}, {{{{{4.4, 5.5, 6.6}}}}}, {{{{{7.7, 8.8, 9.9}}}}}}}}");
   }
+  {
+    ASSERT_EQ(
+      c10::str(torch::detail::TensorDataContainer({{{{{{{{{{1}}}}}}}}}})),
+      "{{{{{{{{{{1}}}}}}}}}}");
+  }
+  {
+    ASSERT_EQ(
+      c10::str(torch::detail::TensorDataContainer({{{{{{{{{{}}}}}}}}}})),
+      "{{{{{{{{{{}}}}}}}}}}");
+  }
+  {
+    ASSERT_EQ(
+      c10::str(torch::detail::TensorDataContainer({{{{{{{{{{1, 2}}}}}}}}}})),
+      "{{{{{{{{{{1, 2}}}}}}}}}}");
+  }
+}
```

Comment:
I don't think you have a test for pretty-printing a TensorDataContainer that holds a tensor

Reply:
Thanks for the catch! Fixed.

Reply:
Added the tests (which caught a printing bug and I just fixed it :D)